### PR TITLE
[ISSUE #790] Knative Connector: Initial Implementation of Producer.

### PR DESCRIPTION
**Pull request by pchengma**
Fixing ISSUE #790.

### Motivation

*Implemented and tested Producer module to publish an event to Knative ([ISSUE #790](https://github.com/apache/incubator-eventmesh/issues/790)).*
*The PRs for [ISSUE #790](https://github.com/apache/incubator-eventmesh/issues/790) will be merged into knative-connector branch.*

### Modifications

*Followed the API design style of [eventmesh-connector-api](https://github.com/apache/incubator-eventmesh/tree/master/eventmesh-connector-plugin/eventmesh-connector-api/src/main/java/org/apache/eventmesh/api) and made the following modifications.*

#### Implementation
- Producer for publishing an event message to Knative
- CloudEvent Reader, Writer, and Header
- Connector
- Unit test for Producer
- Included eventmesh-connector-knative in setting.gradle

#### How to Test

*Assume the cloudevents-player Knative service has already been set up on your local machine; otherwise, please follow the steps in https://knative.dev/docs/getting-started/first-source/#sending-an-event to set up a Knative service (cloudevents-player) as a sink.*

```bash
# Set eventMesh.connector.plugin.type=knative in eventmesh-runtime/conf/eventmesh.properties
# Do unit test:
$ cd incubator-eventmesh/eventmesh-connector-plugin/eventmesh-connector-knative
$ ../../gradlew clean test --tests KnativeConnectorTest.testConnect
# See the delivered event:
$ curl http://cloudevents-player.default.127.0.0.1.sslip.io/messages
```

### Documentation

- Does this pull request introduce a new feature? yes.
- If yes, how is the feature documented? not documented.
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation. The followup issues or pull requests for documentation will be created after almost all implementation work has been finished.

### CODE REVIEW

**Comment by pchengma** [comment]
@xwm1992 Please have a look when you have time. Thanks very much in advance!

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    public KnativeMessageWriter(String data) {
+        String s = "{ \"msg\": [\"" + data + "\"]}";
+        this.message = new CloudEventBuilder()
+                .withId("my-id")
+                .withSource(URI.create("/myClient"))
+                .withType("dev.knative.cronjob.event")
+                .withDataContentType("application/json")
+                .withData(s.getBytes(StandardCharsets.UTF_8))
+                .build();
+    }
```
Comment (xwm1992) [inline eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java:46]: These codes look strange, you use the cloudeventbuilder to build the cloud events, but you only want to put the data, other attributes are useless.

**Review by xwm1992 — CHANGES_REQUESTED** [review]

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/KnativeProducerImpl.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        return producer.isStarted();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return producer.isClosed();
+    }
+
+    @Override
+    public void sendOneway(CloudEvent cloudEvent) {
```
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/KnativeProducerImpl.java:47]: Thanks for your comments. I will change to the ```publish``` method instead.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/AbstractProducer.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        this.httpUrlConnection = (HttpURLConnection) url.openConnection();
+        httpUrlConnection.setDoOutput(true);
+        httpUrlConnection.setDoInput(true);
+
+        // Set HTTP header for CloudEvent:
+        this.httpUrlConnection.setRequestProperty(KnativeHeaders.CONTENT_TYPE, properties.getProperty(KnativeHeaders.CONTENT_TYPE));
+        this.httpUrlConnection.setRequestProperty(KnativeHeaders.CE_ID, properties.getProperty(KnativeHeaders.CE_ID));
+        this.httpUrlConnection.setRequestProperty(KnativeHeaders.CE_SPECVERSION, properties.getProperty(KnativeHeaders.CE_SPECVERSION));
+        this.httpUrlConnection.setRequestProperty(KnativeHeaders.CE_TYPE, properties.getProperty(KnativeHeaders.CE_TYPE));
+        this.httpUrlConnection.setRequestProperty(KnativeHeaders.CE_SOURCE, properties.getProperty(KnativeHeaders.CE_SOURCE));
```
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/AbstractProducer.java:46]: Thanks for your comments. I will move this part out of the ```init``` period.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/ProducerImpl.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                while ((line = reader.readLine()) != null) {
+                    s += line + "\n";
+                }
+                reader.close();
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/ProducerImpl.java:61]: Thanks for your comments. I will change to the asynchronous way to call the ```publish``` method.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java (hunk 1 above)
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java:46]: Thanks for your comments. I will remove the unnecessary attributes.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/KnativeProducerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/KnativeProducerImpl.java]: Implemented ```publish``` method with SendCallback.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/AbstractProducer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/AbstractProducer.java]: Moved HTTP header setting into send() method.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/ProducerImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/producer/ProducerImpl.java]: Implemented ```sendAsync``` method by using AsyncHttpClient.

File: eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java (hunk 1 above)
Comment (pchengma) [reply eventmesh-connector-plugin/eventmesh-connector-knative/src/main/java/org/apache/eventmesh/connector/knative/cloudevent/impl/KnativeMessageWriter.java]: Removed unnecessary attributes in ```KnativeMessageWriter```.

**Pull request merged**